[DAG] visitFREEZE - always allow freezing multiple operands

RKSimon opened the pull request:
Always try to fold freeze(op(....)) -> op(freeze(),freeze(),freeze(),...).

This patch proposes we drop the opt-in limit for opcodes that are allowed to push a freeze through the op to freeze all its operands, through the tree towards the roots.

I'm struggling to find a strong reason for this limit apart from the DAG freeze handling being immature for so long - as we've improved coverage in canCreateUndefOrPoison/isGuaranteedNotToBeUndefOrPoison it looks like the regressions are not as severe.

If there's no objections to this approach I will yak shave some of the remaining regressions.

Hopefully this will help some of the regression issues in #143102 etc.

RKSimon committed:
[DAG] visitFREEZE - always allow freezing multiple operands

Always try to fold freeze(op(....)) -> op(freeze(),freeze(),freeze(),...).

This patch proposes we drop the opt-in limit for opcodes that are allowed to push a freeze through the op to freeze all its operands, bringing us more in line with how InstCombine handles it.

I'm struggling to find a strong reason for this limit apart from the DAG freeze handling being immature for so long - as we've improved coverage in canCreateUndefOrPoison/isGuaranteedNotToBeUndefOrPoison it looks like the regressions are not as severe.

If there's no objections to this approach I will yak shave some of the remaining regressions.

Hopefully this will help some of the regression issues in #143102 etc.

RKSimon committed:
Merge branch 'main' into dag-always-push-freeze

RKSimon committed:
Merge branch 'main' into dag-always-push-freeze

arsenm reviewed (approved):
List looks totally arbitrary so

RKSimon commented:
> List looks totally arbitrary so

Agreed - and it doesn't even seem to be to reduce freeze node counts as ISD::BUILD_VECTOR was allowed to always push and is one of the worst multiplying factors for this kind of thing.

An additional subtlety I've noticed is that a lot of these freezes are now going away entirely as they are not encumbered by the MaxRecursionDepth limit on isGuaranteedNotToBeUndefOrPoison - the freezes just keep getting pushed up until the depth limit is no longer an issue.

I'm trying to address the remaining regressions - they are nearly always missing canCreateUndefOrPoison entries for opcodes: SELECT and the ADD/SUB overflow nodes are the most common generics, or are target specific (e.g. AMDGPUISD FFBH_U32 etc. hinder the i128 div/rem tests)

RKSimon committed:
Merge branch 'main' into dag-always-push-freeze

RKSimon committed:
Update tests now that ISD::SELECT has been added to canCreateUndefOrPoison

RKSimon committed:
Add add/sub/mul overflow opcodes to canCreateUndefOrPoison - both the result and overflow outputs won't create poison

RKSimon committed:
Merge branch 'main' into dag-always-push-freeze

nikic commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -5555,6 +5555,23 @@ bool SelectionDAG::canCreateUndefOrPoison(SDValue Op, const APInt &DemandedElts,
+  case ISD::UADDO_CARRY:
+  case ISD::USUBO_CARRY:
+    // No poison on result or overflow flags.
+    return false;
```
Land this separately?

RKSimon replied:
sure, although I haven't tried to create standalone test cases yet - I'm still hoping to reduce the regressions with further yak shaving.

RKSimon committed:
Merge branch 'upstream' into dag-always-push-freeze

RKSimon committed:
Merge branch 'upstream' into dag-always-push-freeze

RKSimon committed:
[NVPTX] Regenerate i128.ll

RKSimon committed:
Merge branch 'upstream' into dag-always-push-freeze

RKSimon committed:
Merge branch 'upstream' into dag-always-push-freeze

RKSimon committed:
Merge branch 'main' into dag-always-push-freeze

RKSimon committed:
Merge branch 'dag-always-push-freeze' of https://github.com/RKSimon/llvm-project into dag-always-push-freeze

RKSimon committed:
Merge branch 'main' into dag-always-push-freeze

RKSimon committed:
[RISCV] lowerINSERT_SUBVECTOR - fold insert_subvector(freeze(undef),sub,idx) -> insert_subvector(undef,sub,idx) to match existing widening patterns.

RKSimon commented:
Any more comments before I commit?

I've added riscv-specific handling for the insert_subvector(freeze(undef),sub,idx) regression - at this point I think I've cleaned up the notable regressions.

RKSimon committed:
Remove insert_subvector(freeze(undef),sub,idx) -> insert_subvector(undef,sub,idx) hack

RKSimon committed:
[RISCV] lowerVectorFTRUNC_FCEIL_FFLOOR_FROUND - freeze source value before conversion to scalable type

nikic commented on llvm/test/CodeGen/X86/freeze-vector.ll:
```diff
@@ -278,16 +278,22 @@ define void @freeze_buildvector_single_maybe_poison_operand(ptr %origin, ptr %ds
+; X86-NEXT:    movl $42, %ecx
+; X86-NEXT:    vpinsrd $1, %ecx, %xmm0, %xmm0
+; X86-NEXT:    vpinsrd $2, %ecx, %xmm0, %xmm0
+; X86-NEXT:    vpinsrd $3, %ecx, %xmm0, %xmm0
```
What's the cause for this regression? Does this leave behind a freeze poison for the base vector, even though it's non-demanded?

RKSimon committed:
Merge branch 'upstream' into dag-always-push-freeze

RKSimon committed:
[DAG] visitINSERT_VECTOR_ELT - create BUILD_VECTOR from insert chain starting from freeze(undef)

nikic reviewed (approved):
LGTM

RKSimon committed:
Merge branch 'main' into dag-always-push-freeze

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -23205,6 +23196,10 @@ SDValue DAGCombiner::visitINSERT_VECTOR_ELT(SDNode *N) {
 
+      // FREEZE(UNDEF) - build new BUILD_VECTOR from already inserted operands.
+      if (ISD::isFreezeUndef(CurVec.getNode()))
+        return CanonicalizeBuildVector(Ops, /*FreezeUndef=*/true);
```
I can add a OneUse check if that helps? I haven't gotten alive2 to complain about multiple uses:
```
----------------------------------------
declare void @use(<4 x i4>)

define <4 x i4> @src(i4 %a0) {
#0:
  %f = freeze <4 x i4> poison
  call void @use(<4 x i4> %f)
  %v0 = insertelement <4 x i4> %f, i4 %a0, i32 0
  ret <4 x i4> %v0
}
=>
declare void @use(<4 x i4>)

define <4 x i4> @tgt(i4 %a0) {
#0:
  %f = freeze <4 x i4> poison
  call void @use(<4 x i4> %f)
  %f0 = freeze i4 poison
  %v0 = insertelement <4 x i4> poison, i4 %a0, i32 0
  %v1 = insertelement <4 x i4> %v0, i4 %f0, i32 1
  %v2 = insertelement <4 x i4> %v1, i4 %f0, i32 2
  %v3 = insertelement <4 x i4> %v2, i4 %f0, i32 3
  ret <4 x i4> %v3
}
Transformation seems to be correct!
```

RKSimon committed:
[DAG] visitINSERT_VECTOR_ELT - create BUILD_VECTOR from insert chain starting from oneuse(freeze(undef))

RKSimon committed:
Merge branch 'upstream' into dag-always-push-freeze

RKSimon committed:
refresh test

rupprecht commented:
It looks like this causes a crash in certain cases -- `Running pass 'X86 DAG->DAG Instruction Selection' on function`. I assume that's not a known issue yet? I'm trying to reduce.

eaeltsin commented:
@RKSimon - the issue pointed out by @rupprecht is very painful for us. Can you please prioritize the fix? Thank you!

rupprecht commented:
> Sure - to save me time please can you create a llc-only repro?

The provided repro also crashes via `llc -relocation-model=pic -code-model=medium < repro.ll`

https://godbolt.org/z/aMso1cqME

The full debug info may not be necessary to crash, but I couldn't figure out how to reduce that further. The `%load` line also appears not to be necessary.
